File: 백준/Gold/16472. 고냥이/고냥이.py
import sys
input = sys.stdin.readline

n = int(input())
st = input().rstrip()
s, e = 0, 0
dic = {}
ans = 0
while e < len(st):
    dic[st[e]] = dic.get(st[e], 0) + 1
    tmp = len(dic.keys())

    if tmp <= n:
        ans = max(ans, e - s + 1)
        # print(e, s)

    elif tmp > n:
        dic[st[s]] -= 1
        if dic[st[s]] == 0:
            del dic[st[s]]
        s += 1
    e += 1

print(ans)

File: 백준/Gold/16472. 고냥이/고냥이.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        
        int n = Integer.parseInt(br.readLine());
        char[] arr = br.readLine().toCharArray();
        int ans = Integer.MIN_VALUE;

        int[] alpha = new int[26];
        alpha[arr[0] - 'a']++;
        
        int left = 0;
        int right = 0;
        int count = 1;

        while (right < arr.length - 1) {
            right++;

            if (alpha[arr[right] - 'a']++ == 0) count++;

            while (count > n && left < right) {
                if (--alpha[arr[left++] - 'a'] == 0) count--;
            }

            ans = Math.max(ans, right - left + 1);

        }
        System.out.println(ans);
    }
}